src/sentry/incidents/subscription_processor.py

```diff
@@ -0,0 +1,198 @@
+            ):
+                self.trigger_alert_threshold()
+            elif (
+                self.active_incident
+                # TODO: Need to make `resolve_threshold` nullable so that it can be
+                # optional
+                and self.alert_rule.resolve_threshold is not None
+                and resolve_operator(aggregation_value, self.alert_rule.resolve_threshold)
+            ):
+                self.trigger_resolve_threshold()
+            else:
+                self.alert_triggers = 0
+                self.resolve_triggers = 0
+
+            self.update_alert_rule_stats()
+
+    def trigger_alert_threshold(self):
+        """
+        Called when a subscription update exceeds the value defined in the
+        `alert_rule.alert_threshold`, and there is not already an active incident. Increments the
+        count of how many times we've consecutively exceeded the threshold, and if
+        above the `threshold_period` defined in the alert rule then create an incident.
+        :return:
+        """
+        self.alert_triggers += 1
+        if self.alert_triggers >= self.alert_rule.threshold_period:
+            detected_at = to_datetime(self.last_update)
+            self.active_incident = create_incident(
+                self.alert_rule.project.organization,
+                IncidentType.ALERT_TRIGGERED,
+                # TODO: Include more info in name?
+                self.alert_rule.name,
+                alert_rule=self.alert_rule,
+                # TODO: Incidents need to keep track of which metric to display
+                query=self.subscription.query,
+                date_started=detected_at,
+                date_detected=detected_at,
+                projects=[self.alert_rule.project],
+            )
+            # TODO: We should create an audit log, and maybe something that keeps
+            # all of the details available for showing on the incident. Might be a json
+            # blob or w/e? Or might be able to use the audit log.
+
+            # We now set this threshold to 0. We don't need to count it anymore
+            # once we've triggered an incident.
+            self.alert_triggers = 0
+
+    def trigger_resolve_threshold(self):
+        """
+        Called when a subscription update exceeds the value defined in
+        `alert_rule.resolve_threshold` and there's a current active incident.
+        :return:
+        """
+        self.resolve_triggers += 1
+        if self.resolve_triggers >= self.alert_rule.threshold_period:
+            update_incident_status(self.active_incident, IncidentStatus.CLOSED)
+            self.resolve_triggers = 0
+            self.active_incident = None
+
+    @property
+    def alert_rule_stat_keys(self):
+        key_base = self.ALERT_RULE_BASE_STAT_KEY % self.alert_rule.id
+        return [key_base % stat_key for stat_key in self.ALERT_RULE_STAT_KEYS]
+
+    def get_alert_rule_stats(self):
```

Comment:
Is how the redis keys are handled a concern that could be separated out? It seems separate from some of the other logic.

Reply:
They could just be functions on the module I suppose?

Reply:
Actually I'm not totally sure I follow. Are you saying that we shouldn't be saving these stats here at all? Or just that the key names and redis client should be elsewhere.